indigo/analysis/refs/base.py

```diff
@@ -123,3 +123,154 @@ class RefsFinderENG(BaseRefsFinder):
+
+                    # if nothing in the tail matches anymore, we're done
+                    if not matches:
+                        break
+
+                    # there might only be baddies left,
+                    # in which case we're also done
+                    valid_matches = [self.is_valid(match, node) for match in matches]
+                    if True not in valid_matches:
+                        break
+
+                    # mark the reference and continue to check the new tail
+                    node = self.mark_refs_in_matches(matches, node, in_tail=True)
+
+    def mark_refs_in_matches(self, matches, node, in_tail):
+        for match in matches:
+            if self.is_valid(match, node):
+                node = self.mark_reference(node, match, in_tail)
+                break
+
+        return node
+
+    def is_valid(self, match, node):
+        # check that it's not an external reference
+        ref = match.group('ref')
+        if ref.endswith('the ') or ref.endswith('Act '):
+            return False
+
+        # check that it's not a bad reference
+        num = match.group('num')
+        if not node.xpath(f"//a:section[a:num[text()='{num}.']]", namespaces=self.nsmap):
+            return False
+
+        return True
+
+    def mark_reference(self, node, match, in_tail):
+        ref, start_pos, end_pos = self.make_ref(node, match)
+
+        if in_tail:
+            node.addnext(ref)
+            node.tail = match.string[:start_pos]
+            ref.tail = match.string[end_pos:]
+        else:
+            node.text = match.string[:start_pos]
+            node.insert(0, ref)
+            ref.tail = match.string[end_pos:]
+
+        return ref
+
+    def make_ref(self, node, match):
+        """ Make a reference out of this match, returning a (ref, start, end) tuple
+        which is the new ref node, and the start and end position of what text
+        in the parent element it should be replacing.
+        """
+        ref = etree.Element(self.ref_tag)
+        ref.text = match.group('section_ref')
+        ref.set('href', self.make_href(node, match))
+        return ref, match.start('section_ref'), match.end('section_ref')
+
+    def ancestor_nodes(self, root):
+        for x in self.ancestor_xpath(root):
+            yield x
+
+    def candidate_nodes(self, root):
+        for x in self.candidate_xpath(root):
+            yield x
+
+
+@plugins.register('section-refs')
+class SectionRefsFinderENG(BaseSectionRefsFinder):
+    """ Finds internal references to sections in documents, of the form:
+
+        section 26
+        section 26B
+        TODO: match subsections
+        TODO: match paragraphs
+        TODO: match multiple sections
+        TODO: match ranges of sections
+
+    """
+
+    # country, language, locality
+    locale = (None, 'eng', None)
+
+    section_re = re.compile(
+        r'''(?P<ref>
+        (?P<section_ref>
+        \b[sS]ections?\s+
+        (?P<num>\d+(?!\()([A-Z]+)?)
```

Comment:
The negative lookahead for the `(` is redundant, since you know exactly what you want to look for: `\d+[a-z]*`. I suggest making the whole regex case insensitive (`re.IGNORECASE` flag).

Reply:
The new negative lookahead for `(` will discard any section references below subparagraph level, so e.g. `section 26(5)(b)(iv)(A)` won't come up as a match at all. we need to know what comes after the whole reference to check if it's an external reference (and thus ignore it).
I'd rather not make it case insensitive as the casing of the different elements is usually significant.